haystack_experimental/core/pipeline/pipeline.py

```diff
@@ -189,32 +243,52 @@ def run(  # noqa: PLR0915, PLR0912
+            When a breakpoint is triggered. Contains the component name, state, and partial results.
         """
         pipeline_running(self)
+        self.resume_state = resume_state
+        self.debug_path = debug_path
+
+        # make sure breakpoints are valid and have a default visit count
+        validated_breakpoints = self._validate_breakpoints(breakpoints) if breakpoints else None
 
         # TODO: Remove this warmup once we can check reliably whether a component has been warmed up or not
         # As of now it's here to make sure we don't have failing tests that assume warm_up() is called in run()
         self.warm_up()
 
-        # normalize `data`
-        data = self._prepare_component_input_data(data)
-
-        # Raise ValueError if input is malformed in some way
-        self._validate_input(data)
-
         if include_outputs_from is None:
             include_outputs_from = set()
 
-        # We create a list of components in the pipeline sorted by name, so that the algorithm runs deterministically
-        # and independent of insertion order into the pipeline.
-        ordered_component_names = sorted(self.graph.nodes.keys())
+        if not self.resume_state:
+            # normalize `data`
+            data = self._prepare_component_input_data(data)
+
+            # Raise ValueError if input is malformed in some way
+            self._validate_input(data)
+
+            # We create a list of components in the pipeline sorted by name, so that the algorithm runs
+            # deterministically and independent of insertion order into the pipeline.
+            self.ordered_component_names = sorted(self.graph.nodes.keys())
+
+            # We track component visits to decide if a component can run.
+            component_visits = dict.fromkeys(self.ordered_component_names, 0)
 
-        # We track component visits to decide if a component can run.
-        component_visits = dict.fromkeys(ordered_component_names, 0)
+            # We need to access a component's receivers multiple times during a pipeline run.
+            # We store them here for easy access.
+            cached_receivers = {name: self._find_receivers_from(name) for name in self.ordered_component_names}
 
-        # We need to access a component's receivers multiple times during a pipeline run.
-        # We store them here for easy access.
-        cached_receivers = {name: self._find_receivers_from(name) for name in ordered_component_names}
+        else:
+            self._validate_components_state(self.resume_state)
+            data = self._prepare_component_input_data(self.resume_state["pipeline_state"]["inputs"])
+            component_visits = self.resume_state["pipeline_state"]["component_visits"]
+            self.ordered_component_names = self.resume_state["pipeline_state"]["ordered_component_names"]
+            cached_receivers = {name: self._find_receivers_from(name) for name in self.ordered_component_names}
```

Comment:
We're doing this in the `if` case and in `else` case. If we move the log message a few lines up, we can move 
`cached_receivers = {..` out of the if / else.